discovery/sync_manager.go

```diff
@@ -0,0 +1,619 @@
+	select {
+	case m.newActiveSyncers <- &activeSyncer{
+		gossipSyncer: s,
+		transition:   transition,
+	}:
+	case <-m.quit:
+	}
+}
+
+// dequeueActiveSyncer removes the syncer for the given peer from the
+// round-robin queue.
+func (m *syncManager) dequeueActiveSyncer(peer routing.Vertex) {
+	select {
+	case m.staleActiveSyncers <- peer:
+	case <-m.quit:
+	}
+}
+
+// syncerHandler is the syncManager's main event loop responsible for:
+//
+//   1. Finding new peers to receive graph updates from to ensure we don't only
+//   receive them from the same set of peers.
+//
+//   2. Finding new peers to force a full sync with to ensure we have as much of
+//   the public network as possible.
+//
+//   3. Managing the round-robin queue of our active syncers to ensure they
+//   don't overlap and request the same set of channels, which significantly
+//   reduces bandwidth usage.
+//
+// NOTE: This must be run as a goroutine.
+func (m *syncManager) syncerHandler() {
+	defer m.wg.Done()
+
+	// We'll start all of our relevant tickers, except for our
+	// ActiveSyncerTimeoutTicker, which should only be started when we start
+	// an activeSyncer.
+	m.cfg.RotateTicker.Resume()
+	defer m.cfg.RotateTicker.Stop()
+	m.cfg.FullSyncTicker.Resume()
+	defer m.cfg.FullSyncTicker.Stop()
+	defer m.cfg.ActiveSyncerTimeoutTicker.Stop()
+
+	var (
+		// current will hold the current activeSyncer we're waiting for
+		// to complete its state machine.
+		current *activeSyncer
+
+		// startNext will be responsible for containing the signal of
+		// when the current activeSyncer has completed its state
+		// machine. This signal allows us to start the next pending
+		// activeSyncer, if any.
+		startNext chan struct{}
+	)
+
+	// startSyncer is a helper closure we'll use to start a given syncer. If
+	// necessary, a sync transition to an activeSync sync type will be
+	// performed to ensure we receive new graph updates from this peer.
+	startSyncer := func(s *activeSyncer) error {
+		if s.transition {
+			err := s.ProcessSyncTransition(activeSync)
+			if err != nil {
+				return fmt.Errorf("unable to transition to "+
+					"%v: %v", activeSync, err)
+			}
+			s.transition = false
+
+			m.syncersMtx.Lock()
+			m.activeSyncers[s.cfg.peerPub] = s.gossipSyncer
+			delete(m.syncers, s.cfg.peerPub)
+			m.syncersMtx.Unlock()
+		}
+
+		log.Debugf("Starting active syncer for peer=%x", s.cfg.peerPub)
+
+		startNext = s.ResetSyncedSignal()
+		m.cfg.ActiveSyncerTimeoutTicker.Resume()
+		s.Start()
+
+		return nil
+	}
+
+	// startNextSyncer is a helper closure that we'll use to start the next
+	// syncer queued up. If there aren't any, this will act as a NOP.
+	startNextSyncer := func() {
+		current = m.nextPendingActiveSyncer()
+		for current != nil {
+			err := startSyncer(current)
+			if err == nil {
+				return
+			}
+
+			log.Errorf("Unable to start gossipSyncer(%x): %v",
+				current.cfg.peerPub, err)
+
+			current = m.nextPendingActiveSyncer()
+		}
+
```

Comment:
We should log if we get here, as it means that we were unable to start any of the candidate active syncers.

Reply:
Fixed.